// @progress/kendo-react-dateinputs/dist/npm/calendar/components/TodayCommand.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var messages_1 = require("../../messages");
var utils_1 = require("../../utils");
var defaults_1 = require("../../defaults");
var TodayCommand = /** @class */ (function (_super) {
    __extends(TodayCommand, _super);
    function TodayCommand() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleClick = function (syntheticEvent) {
            if (!_this.todayIsInRange) {
                return;
            }
            if (_this.props.onClick) {
                _this.props.onClick.call(undefined, {
                    syntheticEvent: syntheticEvent,
                    nativeEvent: syntheticEvent.nativeEvent,
                    value: utils_1.dateInRange(utils_1.getToday(), _this.min, _this.max),
                    target: _this,
                    isTodayClick: true
                });
            }
        };
        return _this;
    }
    Object.defineProperty(TodayCommand.prototype, "min", {
        get: function () {
            return this.props.min !== undefined
                ? this.props.min
                : TodayCommand.defaultProps.min;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TodayCommand.prototype, "max", {
        get: function () {
            return this.props.max !== undefined
                ? this.props.max
                : TodayCommand.defaultProps.max;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TodayCommand.prototype, "todayIsInRange", {
        get: function () {
            return utils_1.isInRange(utils_1.getToday(), kendo_date_math_1.getDate(this.min), kendo_date_math_1.getDate(this.max));
        },
        enumerable: true,
        configurable: true
    });
    TodayCommand.prototype.render = function () {
        this.localization = kendo_react_intl_1.provideLocalizationService(this);
        var todayMessage = this.localization.toLanguageString(messages_1.today, messages_1.messages[messages_1.today]);
        var todayClassNames = kendo_react_common_1.classNames('k-nav-today', {
            'k-state-disabled': this.props.disabled
        });
        return (React.createElement("span", { className: todayClassNames, onClick: this.handleClick }, todayMessage));
    };
    TodayCommand.propTypes = {
        max: PropTypes.instanceOf(Date).isRequired,
        min: PropTypes.instanceOf(Date).isRequired,
        onClick: PropTypes.func,
        disabled: PropTypes.bool
    };
    TodayCommand.defaultProps = {
        min: defaults_1.MIN_DATE,
        max: defaults_1.MAX_DATE
    };
    return TodayCommand;
}(React.Component));
exports.TodayCommand = TodayCommand;
kendo_react_intl_1.registerForLocalization(TodayCommand);
